app/models/piece.rb

```diff
@@ -98,11 +99,144 @@ def is_on_board?(x,y)
   end
 
   def is_in_check?(x = self.position_x, y = self.position_y)
+    game = self.game
     in_check = false
-    game.pieces.each do |piece|
-      (piece.color != self.color && piece.valid_move?(x,y)) ? in_check = true : in_check = false
+    game.pieces.each do |enemy|
+      if enemy.color != self.color
+        if enemy.valid_move?(x,y)
+         #  puts "[#{x},#{y}]"
+         # puts enemy.inspect
+         in_check = true
+        end
+      end
     end
     return in_check
   end
 
+  def is_in_checkmate?
+    game = self.game
+    in_checkmate = true
+    @squares = []
+    numbers = [1,2,3,4,5,6,7,8]
+    numbers.each do |x|
+      numbers.each do |y|
+        @squares << [x,y]
+      end
+    end
+ 
+    color = self.color
+    king = self
+    in_checkmate = true
+    
+    #if no king is in check there cannot be any checkmate situation
+    if !king.is_in_check?
+      in_checkmate = false
+    else
+      #setting variables, check and store which piece(s) create the check situation
+      x = king.position_x
+      y = king.position_y
+      if color == "black"  
+        attacking_color = "white" 
+      else
+        attacking_color = "black"
+      end
+
+      threatening_pieces = []
+      game.pieces.where(color: attacking_color).each do |piece|
+        if piece.valid_move?(x,y)
+          threatening_pieces << piece
+        end  
+      end
+
+      #check if the king can move to a square where it is not in check.
+      if king.escapable? 
+        in_checkmate = false 
```

Comment:
we could `return false` right away here.. and all the rest will not be evaluated..